sky/packages/sky/lib/src/widgets/heroes.dart

```diff
@@ -0,0 +1,389 @@
+// involve actually scaling and fading, at a minimum.
+
+// Heroes and the Navigator's Stack must be axis-aligned for all this to work.
+// The top left and bottom right coordinates of each animated Hero will be
+// converted to global coordinates and then from there converted to the
+// Navigator Stack's coordinate space, and the entire Hero subtree will, for the
+// duration of the animation, be lifted out of its original place, and
+// positioned on that stack. If the Hero isn't axis aligned, this is going to
+// fail in a rather ugly fashion. Don't rotate your heroes!
+
+// To make the animations look good, it's critical that the widget tree for the
+// hero in both locations be essentially identical. The widget of the target is
+// used to do the transition: when going from route A to route B, route B's
+// hero's widget is placed over route A's hero's widget, and route A's hero is
+// hidden. Then the widget is animated to route B's hero's position, and then
+// the widget is inserted into route B. When going back from B to A, route A's
+// hero's widget is placed over where route B's hero's widget was, and then the
+// animation goes the other way.
+
+// TODO(ianh): If the widgets use Inherited properties, they are taken from the
+// Navigator's position in the widget hierarchy, not the source or target. We
+// should interpolate the inherited properties from their value at the source to
+// their value at the target. See: https://github.com/flutter/engine/issues/1698 
+
+const Object centerOfAttentionHeroTag = const Object();
+
+class _HeroManifest {
+  const _HeroManifest({
+    this.key,
+    this.config,
+    this.sourceStates,
+    this.currentRect,
+    this.currentTurns
+  });
+  final GlobalKey key;
+  final Widget config;
+  final Set<HeroState> sourceStates;
+  final RelativeRect currentRect;
+  final double currentTurns;
+}
+
+abstract class HeroHandle {
+  _HeroManifest _takeChild(Rect animationArea);
+}
+
+class Hero extends StatefulComponent {
+  Hero({
+    Key key,
+    this.navigator,
+    this.tag,
+    this.child,
+    this.turns: 1
+  }) : super(key: key) {
+    assert(tag != null);
+  }
+
+  final NavigatorState navigator;
+  final Object tag;
+  final Widget child;
+  final int turns;
+
+  static Map<Object, HeroHandle> of(BuildContext context, Set<Key> mostValuableKeys) {
+    mostValuableKeys ??= new Set<Key>();
+    assert(!mostValuableKeys.contains(null));
+    // first we collect ALL the heroes, sorted by their tags
+    Map<Object, Map<Key, HeroState>> heroes = <Object, Map<Key, HeroState>>{};
+    void visitor(Element element) {
+      if (element.widget is Hero) {
+        StatefulComponentElement<Hero, HeroState> hero = element;
+        Object tag = hero.state.config.tag;
+        assert(tag != null);
+        Key key = hero.state.config.key;
+        heroes.putIfAbsent(tag, () => <Key, HeroState>{});
+        assert(!heroes[tag].containsKey(key));
+        heroes[tag][key] = hero.state;
```

Comment:
putIfAbsent returns the entry in the map, so you don't have to lookup by `[tag]` again.  You can just index the return value of putIfAbsent.